torch/_inductor/codegen/triton_foreach.py

```diff
@@ -0,0 +1,178 @@
+        with code.indent():
+            if self.block_count == 0:
+                code.splice("pid_offset = pid")
+            else:
+                code.splice(f"pid_offset = pid - {lower_bound_pid}")
+        self.block_count += num_blocks
+
+    def create_sub_kernel(self, *groups, index_dtype, mutations, reduction_hint):
+        sub_kernel = TritonKernel(
+            *groups,
+            index_dtype=index_dtype,
+            mutations=mutations,
+            pid_cache={"tl.program_id(0)": "pid_offset"},
+            reduction_hint=reduction_hint,
+        )
+        metrics.generated_kernel_count -= 1
+        sub_kernel.args = self.args
+        sub_kernel.iter_vars_count = self.iter_vars_count
+        sub_kernel.cse.iter_buffer_ids = self.cse.iter_buffer_ids
+        self.sub_kernels.append(sub_kernel)
+        return sub_kernel
+
+    def jit_line(self):
+        _, _, signature = self.args.python_argdefs()
+        triton_meta = {
+            "signature": dict(enumerate(map(signature_of, signature))),
+            "device": V.graph.scheduler.current_device.index,
+            "constants": {},
+        }
+        triton_meta["configs"] = [config_of(signature)]
+        return (
+            f"@template(num_stages=1, num_warps={self.num_warps}, meta={triton_meta!r})\n"
+            + "@triton.jit"
+        )
+
+    def grid(self):
+        return (
+            self.block_count,
+            1,
+            1,
+        )
+
+    def codegen_kernel(self, name=None):
+        # from triton import next_power_of_2
+
+        code = IndentedBuffer()
+
+        code.splice(
+            """
+                import triton
+                import triton.language as tl
+                from torch._inductor.triton_heuristics import template
+                from torch._inductor.utils import instance_descriptor
+            """
+        )
+        argdefs, _, _ = self.args.python_argdefs()
+        code.writeline(self.jit_line())
+        code.writeline(f"def {name or 'KERNEL_NAME'}({', '.join(argdefs)}):")
+        if config.benchmark_kernel:
+            code.splice(
+                """
+                    from torch._dynamo.testing import rand_strided
+                    from torch._C import _cuda_getCurrentRawStream as get_cuda_stream
+                    import torch
+                    from torch._inductor.triton_heuristics import grid, template
+                    from torch._inductor.utils import instance_descriptor
+                """
+            )
+
+        with code.indent():
+            code.splice("pid = tl.program_id(0)")
+            code.splice(f"XBLOCK: tl.constexpr = {self.block_size}")
+
+            for i in range(next(self.iter_vars_count)):
+                code.splice(f"x{i} = tl.arange(0, XBLOCK)")
+                code.splice("xmask = tl.arange(0, XBLOCK) < XBLOCK")
+
+            for sub_kernel in self.sub_kernels:
+                num_elems = int(sympy_product(sub_kernel.numels))
```

Comment:
So I'll bail in lowering on dynamic shapes for now, I think it should work with dynamic shapes pretty easily but don't need it for the optimizer since the weights are static.

Reply:
Bailed in lowering